### PR TITLE
Beta VAE Training and Model components

### PR DESCRIPTION
**Pull request by jaceybronte**
Reworking the BetaVAE training with best hyperparameters, loss curves, and model component extraction to work with the new Optuna/Pytorch code

### CODE REVIEW

File: 1.data-exploration/scripts/2.test-train_split.py [hunk 1]
```diff
@@ -148,11 +148,44 @@ def save_dataframe(df, file_path: pathlib.Path):
...
+    pat="_", expand=True
+)
+train_and_test_subbed = train_and_test.filter(gene_list_passed_qc, axis=1)
+metadata_holder = []
+metadata_holder = pd.DataFrame(metadata_holder)
+metadata = metadata_holder.assign(
+    ModelID=train_and_test.ModelID.astype(str),
+    AgeCategory=train_and_test.AgeCategory.astype(str),
+    Sex=train_and_test.Sex.astype(str),
+    #train_or_test=train_and_test.train_or_test.astype(str),
```
Comment (MikeLippincott) [inline 1.data-exploration/scripts/2.test-train_split.py]: Consider removing if no longer using

File: 1.data-exploration/scripts/2.test-train_split.py [hunk 2]
```diff
@@ -148,11 +148,44 @@ def save_dataframe(df, file_path: pathlib.Path):
...
+)
+
+metadata_df_dir = pathlib.Path("../0.data-download/data/metadata_df.csv")
+metadata.to_csv(metadata_df_dir, index=False)
+metadata
```
Comment (MikeLippincott) [inline 1.data-exploration/scripts/2.test-train_split.py]: Same here or print the shape and head

File: 1.data-exploration/scripts/2.test-train_split.py [hunk 3]
```diff
@@ -148,11 +148,44 @@ def save_dataframe(df, file_path: pathlib.Path):
...
+
+train_and_test_subbed_dir = pathlib.Path("../0.data-download/data/train_and_test_subbed.csv")
```
Comment (MikeLippincott) [inline 1.data-exploration/scripts/2.test-train_split.py]:
```suggestion
train_and_test_subbed_dir = pathlib.Path("../0.data-download/data/train_and_test_subbed.csv").resolve()
```

File: 1.data-exploration/scripts/2.test-train_split.py [hunk 4]
```diff
@@ -148,11 +148,44 @@ def save_dataframe(df, file_path: pathlib.Path):
...
+gene_dict_df = pd.read_csv("../0.data-download/data/CRISPR_gene_dictionary.tsv", delimiter='\t')
+gene_list_passed_qc = gene_dict_df.query("qc_pass").dependency_column.tolist()
+concat_frames = [train_df, test_df, val_df]
+train_and_test = pd.concat(concat_frames).reset_index(drop=True)
+train_and_test[["AgeCategory", "Sex"]] = train_and_test.age_and_sex.str.split(
+    pat="_", expand=True
+)
+train_and_test_subbed = train_and_test.filter(gene_list_passed_qc, axis=1)
+metadata_holder = []
+metadata_holder = pd.DataFrame(metadata_holder)
```
Comment (MikeLippincott) [inline 1.data-exploration/scripts/2.test-train_split.py]: Consider removing the first assignment of the list as it gets overwritten to a dataframe

File: 1.data-exploration/scripts/2.test-train_split.py [hunk 5]
```diff
@@ -148,11 +148,44 @@ def save_dataframe(df, file_path: pathlib.Path):
...
+
+# In[13]:
+
+
+# create a data frame of both test and train gene effect data with sex, AgeCategory, and ModelID for use in later t-tests
+# load in the data
+
+# create dataframe containing the genes that passed an initial QC (see Pan et al. 2022) and a saturated signal qc, then extracting their corresponding gene label
+gene_dict_df = pd.read_csv("../0.data-download/data/CRISPR_gene_dictionary.tsv", delimiter='\t')
+gene_list_passed_qc = gene_dict_df.query("qc_pass").dependency_column.tolist()
```
Comment (MikeLippincott) [inline 1.data-exploration/scripts/2.test-train_split.py]: Consider using `.loc` instead of `.query` loc is much faster :D

File: 2.train-VAE/betavae.py [hunk 6]
```diff
@@ -72,6 +89,22 @@ def loss_function(self, recon_x, x, mu, log_var):
         KLD = -0.5 * torch.sum(1 + log_var - mu.pow(2) - log_var.exp())
         return MSE + self.beta * KLD
 
+def train_model(model, train_loader, optimizer):
+    """
+    Definition of the VAE training model 
+    """
```
Comment (MikeLippincott) [inline 2.train-VAE/betavae.py]: Consider adding some more documentation to this docstring with type hints too.

```suggestion
def train_model(model, train_loader, optimizer):
    """
    Definition of the VAE training model 
    
    Parameters
    ----------
    model: <replace with type>
        small description of var
    Returns
    -------
    float
        description of float return
    
    """
```

File: 2.train-VAE/betavae.py [hunk 7]
```diff
@@ -72,6 +89,22 @@ def loss_function(self, recon_x, x, mu, log_var):
...
+        optimizer.zero_grad()
+        recon_batch, mu, log_var = model(data)
+        loss = model.loss_function(recon_batch, data, mu, log_var)
+        loss.backward()
+        train_loss += loss.item()
+        optimizer.step()
+    avg_train_loss = train_loss / len(train_loader.dataset)
+    return avg_train_loss
 
 def train_vae(model, train_loader, optimizer, epochs):
```
Comment (MikeLippincott) [inline 2.train-VAE/betavae.py:132]: what is the difference between these two functions?? Consider maybe getting rid of one? or renaming if they carry out different tasks

File: 2.train-VAE/betavae.py [hunk 8]
```diff
@@ -122,3 +145,87 @@ def evaluate_vae(model, val_loader):
...
+        for batch in data_loader:
+            data = batch[0]
+            mu, _ = model.encode(data)
+            latent_space.append(mu.cpu().numpy())
+
+    latent_space = np.concatenate(latent_space, axis=0)
+    latent_df = pd.DataFrame(latent_space)
+    latent_df.insert(0, 'ModelID', metadata['ModelID'])
+
+    latent_df_dir = pathlib.Path("./results/latent_df.csv")
```
Comment (MikeLippincott) [inline 2.train-VAE/betavae.py:222]: Consider making this output path dynamic as it might need to change in the future.

File: 2.train-VAE/betavae.py [hunk 9]
```diff
@@ -122,3 +145,87 @@ def evaluate_vae(model, val_loader):
...
+
+    return latent_df
+
+
+def weights(model, subset_train_df):
+    weight_matrix = model.encoder[0].weight.detach().cpu().numpy().T  # Transpose the weight matrix
+    weight_df = pd.DataFrame(weight_matrix)
+
+    # Save as CSV to use for heatmap
+    weight_df_dir = pathlib.Path("./results/weight_matrix_encoder.csv")
```
Comment (MikeLippincott) [inline 2.train-VAE/betavae.py:243]: same here with the dynamic naming

File: 2.train-VAE/betavae.py [hunk 10]
```diff
@@ -122,3 +145,87 @@ def evaluate_vae(model, val_loader):
...
+    # Remove numbers from gene name column
+
+    split_data_df = gw_renumber_df[0].str.split(" ", expand=True)
+    gene_name_df = split_data_df.iloc[:, :1]
+    trimmed_gene_weight_df = gw_renumber_df.iloc[:, 1:]
+
+    final_gene_weights_df = gene_name_df.join(trimmed_gene_weight_df)
+
+    # Save as CSV to use for GSEA
+    gene_weight_dir = pathlib.Path("./results/weight_matrix_gsea.csv")
```
Comment (MikeLippincott) [inline 2.train-VAE/betavae.py:263]: same here

Comment (MikeLippincott) [inline 2.train-VAE/betavae.py:263]: also consider using `.parquet` files instead :D quicker load and save times

File: 2.train-VAE/betavae.py [hunk 11]
```diff
@@ -122,3 +145,87 @@ def evaluate_vae(model, val_loader):
...
+    latent_df = pd.DataFrame(latent_space)
+    latent_df.insert(0, 'ModelID', metadata['ModelID'])
+
+    latent_df_dir = pathlib.Path("./results/latent_df.csv")
+    latent_df.to_csv(latent_df_dir, index=False)
+
+    return latent_df
+
+
+def weights(model, subset_train_df):
```
Comment (MikeLippincott) [inline 2.train-VAE/betavae.py:228]: please add a doc string :)

File: 2.train-VAE/optimize_utils.py [hunk 12]
```diff
@@ -120,6 +120,8 @@ def get_optimize_args():
 
 
 def objective(trial, train_tensor, val_tensor, train_df):
```
Comment (MikeLippincott) [inline 2.train-VAE/optimize_utils.py:122]: please add a docstring :)

File: 2.train-VAE/scripts/1.optimize_hyperparameters.py [hunk 13]
```diff
@@ -64,3 +69,19 @@
 print(f"Best trial: {best_trial.values}")
 print(f"Best hyperparameters: {best_trial.params}")
 
+
+# In[6]:
+
+
+#Plot and save hyperparameter importance
+save_path = pathlib.Path("../1.data-exploration/figures/param_importance.png")
```
Comment (MikeLippincott) [inline 2.train-VAE/scripts/1.optimize_hyperparameters.py:77]:
```suggestion
save_path = pathlib.Path("../1.data-exploration/figures/param_importance.png").resolve()
```

File: 2.train-VAE/scripts/3.beta_vae_training_extract_model_components.py [hunk 14]
```diff
@@ -4,264 +4,175 @@
...
-
-encoder_architecture = []
-decoder_architecture = []
+# In[3]:
 
 
-# In[9]:
+# Load the best hyperparameters
+study_name = "BetaVAE-Optimization"
+storage_name = f"sqlite:///./{study_name}.db"
```
Comment (MikeLippincott) [inline 2.train-VAE/scripts/3.beta_vae_training_extract_model_components.py:44]: Consider moving this to the results section

File: 1.data-exploration/figures/training_curve_elbow.png
Comment (MikeLippincott) [inline 1.data-exploration/figures/training_curve_elbow.png:1]: Hmmm, the testing and validation data are almost identical?

File: 1.data-exploration/figures/param_importance.png
Comment (MikeLippincott) [inline 1.data-exploration/figures/param_importance.png:1]: Number of layers could be another hyperparameter to assess

**Review by MikeLippincott — APPROVED** [review]
Looks good! Exciting to see the optuna implementation. Most of my comments revolve around documentation of functions, resolving relative paths, and converting csv files to parquets potentially, up to you. Other than that ready to merge after addressing comments! :D

File: 1.data-exploration/scripts/2.test-train_split.py (hunk 2 above)
Comment (jaceybronte) [reply 1.data-exploration/scripts/2.test-train_split.py]: Corrected

File: 1.data-exploration/scripts/2.test-train_split.py (hunk 3 above)
Comment (jaceybronte) [reply 1.data-exploration/scripts/2.test-train_split.py]: Corrected

File: 1.data-exploration/scripts/2.test-train_split.py (hunk 4 above)
Comment (jaceybronte) [reply 1.data-exploration/scripts/2.test-train_split.py]: Corrected

File: 1.data-exploration/scripts/2.test-train_split.py (hunk 5 above)
Comment (jaceybronte) [reply 1.data-exploration/scripts/2.test-train_split.py]: Corrected

File: 2.train-VAE/betavae.py (hunk 6 above)
Comment (jaceybronte) [reply 2.train-VAE/betavae.py]: Corrected

File: 2.train-VAE/betavae.py (hunk 7 above)
Comment (jaceybronte) [reply 2.train-VAE/betavae.py:132]: Need them both separate but clarified in docstring

File: 2.train-VAE/betavae.py (hunk 11 above)
Comment (jaceybronte) [reply 2.train-VAE/betavae.py:228]: Corrected

File: 2.train-VAE/optimize_utils.py (hunk 12 above)
Comment (jaceybronte) [reply 2.train-VAE/optimize_utils.py:122]: Corrected

File: 2.train-VAE/scripts/1.optimize_hyperparameters.py (hunk 13 above)
Comment (jaceybronte) [reply 2.train-VAE/scripts/1.optimize_hyperparameters.py:77]: Corrected
